File: backend/src/prechecks/concrete/node/elastic-version.precheck.ts
import { ElasticClient } from "../../../clients/elastic.client";
import { ConflictError, NotFoundError } from "../../../errors";
import { ClusterNodeType } from "../../../models/cluster-node.model";
import { BaseNodePrecheck } from "../../base/base-node-precheck";
import { ExecutionMode, PrecheckType } from "../../types/enums";
import { NodeContext, PrecheckExecutionRequest } from "../../types/interfaces";

export class ElasticVersionPrecheck extends BaseNodePrecheck {
	private readonly playbookPath: string = "playbooks/pre_checks/elasticsearch-version-check.ansible.yml";
	constructor() {
		super(
			{
				id: "elasticsearch_version_check",
				name: "Elasticsearch Version Check",
				type: PrecheckType.NODE,
				mode: ExecutionMode.CODE,
			},
			ClusterNodeType.ELASTIC
		);
	}

	protected async runForContext(request: PrecheckExecutionRequest<NodeContext>): Promise<void> {
		const currentVersion = request.upgradeJob.currentVersion;
		const client = await ElasticClient.buildClient(request.cluster.clusterId);
		const elasticsearchClient = client.getClient();
		const nodeId = request.context.node.nodeId;
		try {
			const nodes = await elasticsearchClient.nodes.info({
				node_id: nodeId,
			});
			const node = nodes.nodes[nodeId];
			if (node.version === currentVersion) {
				await this.addLog(request, `Node is running on the expected version: ${currentVersion}.`);
			} else {
				const message = `Node version mismatch: expected ${currentVersion}, but found ${node.version}.`;
				throw new ConflictError(message);
			}
		} catch (err) {
			const message = `Node with ID ${nodeId} not found`;
			throw new NotFoundError(message);
		}
	}
}

File: server/src/main/java/co/hyperflex/prechecks/concrete/node/elastic/ElasticVersionPrecheck.java
package co.hyperflex.prechecks.concrete.node.elastic;

import co.elastic.clients.elasticsearch.ElasticsearchClient;
import co.elastic.clients.elasticsearch.nodes.NodesInfoResponse;
import co.elastic.clients.elasticsearch.nodes.info.NodeInfo;
import co.hyperflex.prechecks.contexts.NodeContext;
import co.hyperflex.prechecks.core.BaseElasticNodePrecheck;
import java.io.IOException;
import java.util.Map;
import org.slf4j.Logger;
import org.springframework.stereotype.Component;

@Component
public class ElasticVersionPrecheck extends BaseElasticNodePrecheck {


  public ElasticVersionPrecheck() {
  }

  @Override
  public String getName() {
    return "Elasticsearch Version Check";
  }

  @Override
  public void run(NodeContext context) {
    String nodeId = context.getNode().getId();

    String expectedVersion = context.getClusterUpgradeJob().getCurrentVersion();
    ElasticsearchClient client = context.getElasticClient().getElasticsearchClient();
    Logger logger = context.getLogger();

    try {
      NodesInfoResponse response = client.nodes().info(r -> r.nodeId(nodeId));
      Map<String, NodeInfo> nodes = response.nodes();
      NodeInfo node = nodes.get(nodeId);

      if (node == null) {
        throw new RuntimeException("Node with ID [" + nodeId + "] not found.");
      }

      String actualVersion = node.version();
      if (expectedVersion.equals(actualVersion)) {
        logger.info("Node [{}] is running on the expected version: {}.", node.name(),
            expectedVersion);
      } else {
        logger.error("Node [{}] version mismatch: expected {}, but found {}.",
            node.name(), expectedVersion, actualVersion);
        throw new RuntimeException();
      }

    } catch (IOException e) {
      logger.error("Failed to get version info for node: {}", nodeId, e);
      throw new RuntimeException("Failed to get version info for node: " + nodeId, e);
    }
  }
}
